### PR TITLE
ODSs now keep track of their parent and their location in the tree is evaluated dynamically

### LINKED ISSUE #134: Coordinate interpolation always uses coordinate from source 0
If the different sources in `ods.core_sources` have different lengths, there are issues with coordinate interpolation. It looks like all the sources look to `core_sources.source[0].profiles_1d[0].grid.rho_tor_norm` to know if they should interpolate and then it's not done right. 
I made an example here: `/fusion/projects/omfit-results/lyonsbc/projects/ods_source-coord-interp-fail.zip`
The ODS has two sources. `0` has length 201 and `1` has length 51. If I try to interpolate on `core_sources.source[1].profiles_1d[0].grid.rho_tor_norm`, nothing happens. If I try on `core_sources.source[0].profiles_1d[0].grid.rho_tor_norm`, the source `0` interpolation works properly, but then `1` throws an error. Script is in `Command box 1`, results are below. Note that there's a bug in the exception statement, but that's not the problem. The code for the exception is 
```
if len(ods_coordinates.__getitem__(coordinate, None)) != len(value):
    raise Exception(
        'coordsio %s.shape=%s does not match %s.shape=%s'
        % (coordinate, output_coordinates.__getitem__(coordinate, False).shape, location, value.shape)
    )
```
The `output_coordinates` in the `Exception` should be `ods_coordinates`, I think.

```
Original ODS source lengths
0:  201
1:  51

Interpolate Source 1 rho to 101
Should give 201 & 101, but Source 1 does not interpolate
0:  201
1:  51

Interpolate Source 0 rho to 101
Should give 101 & 51, but Source 1 throws error
0:  101
Error in "OMFIT command box #1" at line  22
    print('1: ', len(ods['core_sources.source[1].profiles_1d[0].grid.volume']))
  [Previous line repeated 3 more times]

Exception: coordsio core_sources.source.0.profiles_1d.0.grid.rho_tor_norm.shape=(101,) does not match core_sources.source.0.profiles_1d.0.grid.volume.shape=(51,)


Press <F6> to see full error report...

Traceback (most recent call last):
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/classes/OMFITx.py", line 4068, in manage_user_errors
    tmp=command(**kw)
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/utils_tk.py", line 1060, in GlobLoc_tk
    return py.run(_relLoc=self.namespace, _OMFITscriptsDict=False, _OMFITconsoleDict=True, noGUI=None)
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/classes/omfit_python.py", line 1082, in run
    result = self.__run__(**kw)
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/classes/omfit_python.py", line 931, in __run__
    self, userDict=kw, inputDict=_relLoc, persistentDict=OMFITconsoleDict, runDict={}, prerun=prerun, postrun=postrun
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/classes/omfit_python.py", line 129, in f_locked
    return f(*args, **kw)
  File "/fusion/projects/omfit-results/lyonsbc/OMFIT-source/omfit/classes/omfit_python.py", line 362, in execGlobLoc
    exec(compile(execString, filename, "exec"), GlobLoc)
  File "/local-scratch/lyonsbc/OMFIT/OMFIT_2020-10-26_18_37_01_292295/project/OMFIT command box #1", line 22, in <module>
    print('1: ', len(ods['core_sources.source[1].profiles_1d[0].grid.volume']))
  File "/fusion/projects/omfit-results/lyonsbc/omas/omas/omas_core.py", line 1152, in __getitem__
    return value.__getitem__(key[1:], cocos_and_coords)
  File "/fusion/projects/omfit-results/lyonsbc/omas/omas/omas_core.py", line 1152, in __getitem__
    return value.__getitem__(key[1:], cocos_and_coords)
  File "/fusion/projects/omfit-results/lyonsbc/omas/omas/omas_core.py", line 1152, in __getitem__
    return value.__getitem__(key[1:], cocos_and_coords)
  [Previous line repeated 3 more times]
  File "/fusion/projects/omfit-results/lyonsbc/omas/omas/omas_core.py", line 1215, in __getitem__
    % (coordinate, output_coordinates.__getitem__(coordinate, False).shape, location, value.shape)
Exception: coordsio core_sources.source.0.profiles_1d.0.grid.rho_tor_norm.shape=(101,) does not match core_sources.source.0.profiles_1d.0.grid.volume.shape=(51,)
```

### PR DESCRIPTION
**Pull request by orso82**
close #134

### CODE REVIEW

**Comment by orso82** [comment]
@bclyons12 can you please test your STEP workflow with this branch of OMAS?

File: omas/tests/test_omas_core.py [hunk 1]
```diff
@@ -127,6 +127,32 @@ def test_misc(self):
         )
         return
 
+    def test_dynammic_location(self):
```
Comment (smithsp) [inline omas/tests/test_omas_core.py:130]:
```suggestion
    def test_dynamic_location(self):
```

File: omas/tests/test_omas_core.py [hunk 2]
```diff
@@ -127,6 +127,32 @@ def test_misc(self):
...
+        ods = ODS()
+        for k in range(5):
+            ods[f'equilibrium.time_slice.{k}.global_quantities.ip'] = k
+        tmp = ods[f'equilibrium.time_slice.-1']
+        assert tmp.location == f'equilibrium.time_slice.{k}'
+        del ods[f'equilibrium.time_slice.0']
+        assert tmp.location == f'equilibrium.time_slice.{k-1}'
+
+    def test_auto_deepcopy_on_assignement(self):
```
Comment (smithsp) [inline omas/tests/test_omas_core.py:139]:
```suggestion
    def test_auto_deepcopy_on_assignment(self):
```

File: omas/tests/test_omas_core.py [hunk 3]
```diff
@@ -127,6 +127,32 @@ def test_misc(self):
...
+        ods = ODS()
+        ods[f'equilibrium.time_slice.0.global_quantities.ip'] = 0.0
+        ods[f'equilibrium.time_slice.1'] = ods[f'equilibrium.time_slice.0']
+
+        # test auto copy.deepcopy on assignement
```
Comment (smithsp) [inline omas/tests/test_omas_core.py:144]:
```suggestion
        # test auto copy.deepcopy on assignment
```

File: omas/tests/test_omas_core.py [hunk 4]
```diff
@@ -127,6 +127,32 @@ def test_misc(self):
...
+        assert id(ods[f'equilibrium.time_slice.0']) != id(ods[f'equilibrium.time_slice.1'])
+
+        # test no extra copy if the user does it for us
+        tmp = copy.deepcopy(ods[f'equilibrium.time_slice.0'])
+        ods[f'equilibrium.time_slice.2'] = tmp
+        assert id(ods[f'equilibrium.time_slice.2']) == id(tmp)
+
+        # test extra copy if multiple assignements
```
Comment (smithsp) [inline omas/tests/test_omas_core.py:152]:
```suggestion
        # test extra copy if multiple assignments
```

**Review by smithsp — COMMENTED** [review]
Caught a few misspellings.

File: omas/omas_core.py [hunk 5]
```diff
@@ -952,7 +970,7 @@ def __setitem__(self, key, value):
                         raise IndexError('`%s[%d]` but ods has no data' % (self.location, key[0]))
                     else:
                         raise IndexError(
-                            '`%s[%d]` but maximun index is %d.\nPerhaps you want to set ods.dynamic_path_creation=\'dynamic_array_structures\''
+                            '`%s[%d]` but maximun index is %d\nPerhaps you want to set ods.dynamic_path_creation=\'dynamic_array_structures\''
```
Comment (smithsp) [inline omas/omas_core.py:973]:
```suggestion
                            '`%s[%d]` but maximum index is %d\nPerhaps you want to set ods.dynamic_path_creation=\'dynamic_array_structures\''
```

**Review by bclyons12 — APPROVED** [review]
This appears to work for the STEP workflow

**Pull request merged**